ext/opcache/ZendAccelerator.c

```diff
@@ -2987,6 +2988,8 @@ static int accel_remap_huge_pages(void *start, size_t size, size_t real_size, co
 #  endif
 	}
 
+	zend_mmap_set_name(start, size, "zend_huge_code_pages");
+
 	if (ret == start) {
```

Comment:
Correct me if I'm wrong, but isn't this condition always true given the `MAP_FIXED` flag? Or do we support OSes that don't support this flag? If the check isn't redundant we should use `ret` instead of `start` (and maybe also just set it if it succeeded as otherwise it's not an actual huge page). But if that were the case we'd be passing the misaligned address to `munmap` anyway which would likely already be incorrect, so unlikely. Note that the return value of `accel_remap_huge_pages` is not even checked either.

Reply:
This function is very obscure with confusing `#ifdefs`, and the whole huge code pages feature doesn't make sense (it slows down everything, instead of speeding it up, but that's not this PR's point).
But true, this check shouldn't ever fail (unless the kernel is buggy).
Though it doesn't matter if you use `ret` or `start`.

Reply:
Isn't the main point of remapping to huge tables to avoid TLB misses? So whether this slows things down or speed things up depends on how the size of the block, as well as the way and how frequently it is accessed which can't easily be judged without benchmarking. What doesn't really make sense is that we do anything at all when `MAP_HUGETLB` is not defined.

Reply:
>  Though it doesn't matter if you use ret or start.

That's true. I'm resolving this for now, we can improve in a separate PR.

Reply:
> Isn't the main point of remapping to huge tables to avoid TLB misses? So whether this slows things down or speed things up depends on how the size of the block, as well as the way and how frequently it is accessed which can't easily be judged without benchmarking.

True; in the initial commit 669f0b39b184593e01e677360fd79b2b63058ca0, there was an alleged "2% improvement on real-life apps", with no mention how this measurement was made, what was actually measured and what those "real-life apps" were.
Was CPU time measured or only TLB misses? If the latter, did those reduced TLB misses actually lead to a performance improvement?
Small improvements like 2% can hardly be measured at all, benchmarking noise is usually more than that, and while huge pages do improve TLB misses, I doubt an improvement through huge pages can be that large.

I doubt that there is really an improvement because this allocates additional memory for data which already exists in the page cache. This may be an improvement in microbenchmarks on small servers where there is just one PHP process (and not much else); but since each PHP process will have its own independent copy of those code pages, you increase memory usage, reduce instruction cache effectiveness.

But of course you are right that this should be benchmarked. What I do is just speculation.

> _Edit_: Oh actually we also have the `MADV_HUGEPAGE` call, but I think the temporary copy would be useless in that case (unsure on that point, I'll check the documentation on my next PHP working day).

That wouldn't work. That's the problem (why this "feature" was invented): the (Linux's) page cache can't do huge pages, no matter what you tell it via `madvise()`. There is work underway to change that, has been for many years (https://lwn.net/Articles/686690/), and with the advent of folios in 5.16, this may finally become possible really soon.